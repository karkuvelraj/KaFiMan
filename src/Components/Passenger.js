import React, { useEffect, useState } from "react"
import PropTypes from "prop-types"
import { makeStyles } from '@material-ui/core/styles';
import Modal from '@material-ui/core/Modal';


function getModalStyle() {
  const top = 50 
  const left = 50

  return {
    top: `${top}%`,
    left: `${left}%`,
    transform: `translate(-${top}%, -${left}%)`,
  };
}

const useStyles = makeStyles((theme) => ({
  paper: {
    position: 'absolute',
    width: '50%',
    height:'70%',
    overflow:'auto',
    backgroundColor: theme.palette.background.paper,
    border: '2px solid #000',
    boxShadow: theme.shadows[5],
    padding: theme.spacing(2, 4, 3),
  },
}));

export default function Passenger({passenger,onClose,open,state,save}) {

  const classes = useStyles();

  useEffect(()=>{
    if(passenger.id)
    setTemp({...passenger})
  },[open]) 

 
  const [passengerTemp,setTemp]=useState({...passenger})
  const [service,setService] = useState(false)
  const [shopItems,setShopItems] = useState(false)
  const [special,setSpecial] = useState(false)
  const [modalStyle] = useState(getModalStyle);

  const iterateAndAppend=(array,dataMap)=>{
    return array? array.map((id,i)=>{
        return <tr key={i}><td colSpan="2">{dataMap[id]?dataMap[id].name:""}</td></tr>
    }):null
  }

  const optionsMap=(ids,map)=>{
    let options=[]
    options.push(<option key={999} disabled value=''>Select</option>)
    for(let id in map){
      if(!ids.includes(id))
        options.push(<option key={id} value={id}>{map[id].name}</option>)
    }
    return options
  }

  const changeHandle=(key,array,event,func)=>{
    let values=[...array]
    values.push(event.target.value)
    setTemp({...passengerTemp,[key]:values})
    func(false)
  }
  const removedPref=()=>{
    setTemp({...passengerTemp,mealPrefer:''})
  }
  const body = (
    <div style={modalStyle} className={classes.paper}>
        <span className="heading">Passenger Details</span>
        <table className="outertable">
        <tbody>
            <tr><td>Name </td><td>  {passengerTemp.name}</td></tr>
            <tr><td>Passport No</td><td> {passengerTemp.passportNo}</td></tr>
            <tr><td>Address</td><td> {passengerTemp.address}</td></tr>
            <tr><td>Date of Birth</td><td> {passengerTemp.dob}</td></tr>
            <tr><td colSpan="2">
        <table className="innertable">
            <thead><tr><th>Ancillary Services</th><th>
            {service?<>
            <select  defaultValue=''  id="serviceId" onChange={(e)=>{changeHandle('services',passengerTemp.services,e,setService)}}>
            {optionsMap(passengerTemp.services,state.servicesMap)}
            </select> <button className="r-btn" onClick={()=>setService(false)}>cancel</button></>
            : <button className="p-btn" onClick={()=>setService(true)}>Add</button> }

            </th></tr></thead> 
            <tbody>{iterateAndAppend(passengerTemp.services,state.servicesMap)}</tbody>
         </table>
        </td></tr>
     <tr><td colSpan="2">
        <table className="innertable">
          <thead><tr><th>Special Meal</th><th>
          {special?<>
          <select  defaultValue=''  id="specialId" onChange={(e)=>{setTemp({...passengerTemp,mealPrefer:e.target.value});setShopItems(false)}}>
           {optionsMap([passengerTemp.mealPrefer],state.specialMealsMap)}
          </select>
            <button className="r-btn" onClick={()=>setSpecial(false)}>cancel</button></>
          : <><button  className="p-btn" onClick={()=>setSpecial(true)}>Add Special Meal</button> 
          <button  className="r-btn" onClick={()=>removedPref()}>Remove Meal Preference</button> </>}
           </th></tr>
           </thead>
       <tbody>
       {passengerTemp.mealPrefer?iterateAndAppend([passengerTemp.mealPrefer],state.specialMealsMap):
       <tr ><td colSpan="2">No Meals Preference</td></tr>}
         </tbody></table>
        </td></tr>
     <tr><td colSpan="2">
            <table className="innertable">
            <thead><tr><th>Shop Items </th><th>
          {shopItems?<>
          <select  defaultValue=''  id="shopItemId" onChange={(e)=>{changeHandle('shopItems',passengerTemp.shopItems,e,setShopItems)}}>
           {optionsMap(passengerTemp.shopItems,state.shopItemsMap)}
          </select> <button className="r-btn" onClick={()=>setShopItems(false)}>cancel</button></>:
           <button className="p-btn" onClick={()=>setShopItems(true)}>Add</button> }
            </th></tr>
            </thead>
        <tbody>
        {iterateAndAppend(passengerTemp.shopItems,state.shopItemsMap)}
          </tbody></table>
        </td></tr>
        <tr><td colSpan="2">
          <button className="s-btn" onClick={()=>save(passengerTemp)}>Save</button>
          <button id="modalClose" className="r-btn"onClick={()=>onClose(false)}>Close</button>
        </td></tr>
        </tbody></table>
    </div>
  );

  return (
    <div>
      <Modal
        open={open}
        onClose={()=>onClose(false)}>
        {body}
      </Modal>
    </div>
  );
}
Passenger.propTypes={
  onClose:PropTypes.func.isRequired,
  save:PropTypes.func.isRequired,
  open:PropTypes.bool.isRequired,
  passenger:PropTypes.object.isRequired,
  state:PropTypes.object.isRequired
}
