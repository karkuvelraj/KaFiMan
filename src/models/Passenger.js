class Passenger {
    name=""
    dob=""
    passportNo=""
    seatNo=""
    flightNo=""
    address=""
    isCheckedIn=""
    isWheelChair=""
    isInfant=""
    services=[]
    mealPrefer=""
    shopItems=[]
    

}

export default Passenger;