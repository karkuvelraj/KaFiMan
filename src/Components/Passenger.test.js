import React from 'react'
import {mount, shallow} from 'enzyme'
import Passenger from './Passenger'
import data from '../../tools/db.json'
let passenger= {
    id:"ASD1000",
    name: "Vimal raja",
    dob: "1996-04-05",
    passportNo: "ASD1000",
    seatNo: "4",
    flightNo:"F10W10",
    address: "",
    isCheckedIn: true,
    isWheelChair: false,
    isInfant: false,
    services: [1,2],
    mealPrefer: "",
    shopItems: [2,1,4]
  }
  let props={
    onClose:jest.fn(),
    save:jest.fn(),
    open:true,
    passenger:passenger,
     state:load()
  }
  function load (){
    return {
        shopItemsMap:data.shopItems.reduce((a,b)=>{ a[b.id]=b;return a},{}),
        specialMealsMap:data.specialMeals.reduce((a,b)=>{ a[b.id]=b;return a},{}),
        servicesMap:data.services.reduce((a,b)=>{ a[b.id]=b;return a},{})
    }
    
  }

it('should be opening modal and display passenger',()=>{
    const passengerModal=shallow(<Passenger {...props}/>)
    expect(passengerModal).toMatchSnapshot()
})

it('should display all enabled services of passenger ',()=>{
    const passengerModal=shallow(<Passenger {...props}/>)
    let tbody=passengerModal.find('.innertable').at(0).find('tbody')
    expect(tbody.find('tr').length).toBe(2)
})
it('should display all enabled Shop Items of passenger ',()=>{
    const passengerModal=shallow(<Passenger {...props}/>)
    let tbody=passengerModal.find('.innertable').at(2).find('tbody')
    expect(tbody.find('tr').length).toBe(3)
})

it('should display no meals if no special meals prefered ',()=>{
    const passengerModal=shallow(<Passenger {...props}/>)
    let tbody=passengerModal.find('.innertable').at(1).find('tbody')
    expect(tbody.find('tr').first().find('td').text()).toBe('No Meals Preference')
})

it('should display Meal name of meal prefered by passenger',()=>{
    props.passenger.mealPrefer='BBML'
    const name="INFANT/BABY FOOD"
    const passengerModal=shallow(<Passenger {...props}/>)
    let tbody=passengerModal.find('.innertable').at(1).find('tbody')
    expect(tbody.find('tr').first().find('td').text()).toBe(name)
})

it('should close modal on clicking close button',()=>{
  
    const passengerModal=shallow(<Passenger {...props}/>)
 
    const close=passengerModal.find('#modalClose')
    expect(close.text()).toBe('Close')
    close.simulate('click')
    expect(props.onClose).toBeCalledWith(false)
})


it('should be able to add service ',()=>{

    const passengerModal=mount(<Passenger {...props}/>)
    const table=passengerModal.find('.innertable').at(0);
    let thead=table.find('thead')
    const button=thead.find('tr').first().find('th').at(1).find('button')
    expect(button.text()).toBe('Add')
    button.simulate('click')
    passengerModal.update()
    
    passengerModal.find('select').simulate('change',{target:{value:"3"}})
    passengerModal.update()
    console.log(table.find('tbody').html())
    expect(passengerModal.find('.innertable').at(0)
        .find('tbody').find('tr').getElements().length).toBe(3)
})