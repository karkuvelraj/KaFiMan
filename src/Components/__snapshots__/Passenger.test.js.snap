// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be opening modal and display passenger 1`] = `ShallowWrapper {}`;
